x/cash/dynamicfee.go

```diff
@@ -0,0 +1,235 @@
+	return DynamicFeeDecorator{
+		auth:    auth,
+		control: control,
+	}
+}
+
+// Check verifies and deducts fees before calling down the stack
+func (d DynamicFeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Checker) (weave.CheckResult, error) {
+
+	// prepare fees for subtransaction -- shared in Check and Deliver
+	var res weave.CheckResult
+	fee, payer, cache, err := d.prepare(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	// read config
+	minFee := d.getMinFee(store)
+
+	// do subtransaction in a function for easier error handling
+	res, err = func() (weave.CheckResult, error) {
+		// shadow with local variables...
+		err := d.maybeTakeFee(cache, payer, fee)
+		if err != nil {
+			return weave.CheckResult{}, err
+		}
+		res, err := next.Check(ctx, cache, tx)
+		// TODO: check RequiredFee here and return an error if insufficient
+		return res, err
+	}()
+
+	// on error, rollback, then take the minfee from the store
+	if err != nil {
+		cache.Discard()
+		// if this fails, we aborted early above, we can just ignore return value
+		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
+		d.maybeTakeFee(store, payer, minFee)
+		// return error from the transaction, not the possible error from minFee deduction
+		return res, err
+	}
+
+	// if success, we commit and update the importance
+	cache.Write()
+	res.GasPayment += toPayment(fee)
+	return res, err
+}
+
+// Deliver verifies and deducts fees before calling down the stack
+func (d DynamicFeeDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Deliverer) (weave.DeliverResult, error) {
+
+	var res weave.DeliverResult
+	fee, payer, cache, err := d.prepare(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	// read config
+	minFee := d.getMinFee(store)
+
+	// do subtransaction in a function for easier error handling
+	res, err = func() (weave.DeliverResult, error) {
+		// shadow with local variables...
+		err := d.maybeTakeFee(cache, payer, fee)
+		if err != nil {
+			return weave.DeliverResult{}, err
+		}
+		res, err := next.Deliver(ctx, cache, tx)
+		// TODO: check RequiredFee here and return an error if insufficient
+		return res, err
+	}()
+
+	// on error, rollback, then take the minfee from the store
+	if err != nil {
+		cache.Discard()
+		// if this fails, we aborted early above, we can just ignore return value
+		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
+		d.maybeTakeFee(store, payer, minFee)
+		// return error from the transaction, not the possible error from minFee deduction
+		return res, err
+	}
+
+	// if success, we commit and update the importance
+	cache.Write()
+	return res, err
+}
+
+// maybeTakeFee will send fees only if they are positive, so we don't have to check IsZero() everywhere else
+func (d DynamicFeeDecorator) maybeTakeFee(store weave.KVStore, src weave.Address, amount x.Coin) error {
```

Comment:
I think the name could better reflect what the method does. How about just `takeFee` or `deductFee`?
Having `maybe` in the name does not explain anything. Comment makes it clear, but then the `maybe` is not needed.

Isn't negative fee a configuration issue? Should we at least log such case instead of ignoring it?

Reply:
It was for a zero fee handling. Negative is illegal.

Maybe just update the MoveCoin method to be a noop when moving 0, rather than returning errors

Reply:
> Isn't a negative fee a configuration issue? Should we at least log such case instead of ignoring it?

Yes, we can make this check just for `.IsZero()`. Negative case must be handled elsewhere (both in validation, and in MoveCoin returning an error)